### PR TITLE
OPECO-2596: semver should generate replaces across Y-streams in minor-channel mode

### LINKED ISSUE #1031: opm alpha render-veneer semver missing replaces between minor channels
**Context**

I am using the `opm alpha render-veneer semver` command to generate a catalog from the following veneer:
```yaml
Schema: olm.semver
GenerateMajorChannels: false
GenerateMinorChannels: true
Candidate:
  Bundles: []
Fast:
  Bundles: []
Stable:
  Bundles:
  - Image: docker.io/example/foo:v0.1.0
  - Image: docker.io/example/foo:v0.1.1
  - Image: docker.io/example/foo:v0.2.0
  - Image: docker.io/example/foo:v1.0.0
```
Note that the actual image names were replaced. The bundles are generated using the following command (no default channel set):
```bash
generate bundle -q --overwrite --package=$(PACKAGE_NAME) --version $(VERSION) ...
```

The output is the following:
```yaml
---
defaultChannel: stable-v1.0
name: foo
schema: olm.package
---
entries:
- name: foo.v0.1.0
- name: foo.v0.1.1
  skips:
  - foo.v0.1.0
name: stable-v0.1
package: foo
schema: olm.channel
---
entries:
- name: foo.v0.2.0
name: stable-v0.2
package: foo
schema: olm.channel
---
entries:
- name: foo.v1.0.0
name: stable-v1.0
package: foo
schema: olm.channel
---
```

Versions:
```bash
> opm version
Version: version.Version{OpmVersion:"v1.26.2", GitCommit:"0e53cafe", BuildDate:"2022-09-21T15:25:25Z", GoOs:"darwin", GoArch:"arm64"}

> operator-sdk version
operator-sdk version: "v1.24.0", commit: "de6a14d03de3c36dcc9de3891af788b49d15f0f3", kubernetes version: "1.24.2", go version: "go1.18.6", GOOS: "darwin", GOARCH: "amd64"

```

**Problem**

As you can see in the output above, there is no `replaces` directive between minor channels. I am expecting e.g. `stable-v0.2` to have:
```yaml
entries:
- name: foo.v0.2.0
  replaces: foo.v0.1.1. # <-- this is missing
name: stable-v0.2
package: foo
schema: olm.channel
```

This `replaces` is however present [in the docs](https://olm.operatorframework.io/docs/reference/veneers/):
```yaml
entries:
  - name: testoperator.v1.1.0
    replaces: testoperator.v1.0.1
name: candidate-v1.1
package: testoperator
schema: olm.channel
```

Without this directive, when the `Subscription` is updated from `stable-v0.1` to. `stable-v0.2` (and the operator is installed at version 0.1.1), OLM doesn't see nor performs the upgrade. The only way is to uninstall and then reinstall.

I understood that this is a feature for major channels, as hinted by:

> Please note that at no time do we transgress across major-version boundaries with the channels, to be consistent with [the semver convention](https://semver.org/) for major versions, where the purpose is to make incompatible API changes.

But for minor versions the upgrade should be automatic when switching channels.

### PR DESCRIPTION
**Pull request by theishshah**
This PR contains the initial logic and docs for cleaning up the FBC Z-stream replacements. Follow up work for this PR will include a unit testing deep dive spike and some re-factoring to pre-sort and filter versions.

Closes #1031 

streamlining & add'l namechanges to working utest, debug printfs removed
working idempotent approach across set of all channel entries; but gaps in utest and UGLY implementation refactor minimally-nested ifs
updated supporting docs
standardize on yaml.UnmarshallStrict to give better diagnostics for unexpected fields

### CODE REVIEW

File: alpha/declcfg/write.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ import (
 type MermaidWriter struct {
 	MinEdgeName          string
 	SpecifiedPackageName string
+	CrossChannelLinking  bool
```
Comment (grokspawn) [inline alpha/declcfg/write.go:21]: I forgot that I left the cross-channel linking support in here.  It's not getting used by anything, but it ... also doesn't work. 
I'm open to leaving it as-is or breaking it out.

**Comment by grokspawn** [comment]
Would you please retitle to something like `OPECO-2596: semver should generate replaces across Y-streams in minor-channel mode` and add solves #1031  ?

**Review by everettraven — CHANGES_REQUESTED** [review]
Overall looks pretty good, great work @theishshah!

The reason for me marking this review as requesting changes is some minor things I found in how the tests are written that I *think* could be contributing to the test failures.

Aside from that I had a few smaller nits.

File: alpha/veneer/composite/builder.go [hunk 2]
```diff
@@ -55,7 +55,7 @@ func (bb *BasicBuilder) Build(dir string, vd VeneerDefinition) error {
 	}
 	// Parse out the basic veneer configuration
 	basicConfig := &BasicVeneerConfig{}
-	err := json.Unmarshal(vd.Config, basicConfig)
+	err := yaml.UnmarshalStrict(vd.Config, basicConfig)
```
[the remark this answers is not in the document]
Comment (grokspawn) [reply alpha/veneer/composite/builder.go:58]: This comes back to a comment I put in the composite template origination PR where I was asking for strict field unmarshalling as well as validation of a minimum set of fields (https://github.com/operator-framework/operator-registry/pull/1054#discussion_r1074039456). 

Well, I got distracted and didn't follow up on it, and this part of it didn't happen.   This is a functionally equivalent approach which has the added benefit of disallowing unknown/unexpected members (https://pkg.go.dev/gopkg.in/yaml.v2#UnmarshalStrict). 

This is totally outside the original purview of this effort, and really just a "I found a clever way to cover this base".

File: alpha/declcfg/write.go (hunk 1 above)
Comment (grokspawn) [reply alpha/declcfg/write.go]: Removing these entirely.  I had hoped that we'd be able to use this to visually debug the fact that we now can have `skips` in a channel which are not part of the channel (but within the Y range of the channel). 
But alas, it was just a bit too much effort for the time available.

File: alpha/veneer/semver/types.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	parent  string
+	index   int
+	version semver.Version
+}
+
+func (t entryTuple) String() string {
+	return fmt.Sprintf("{ arch: %q, kind: %q, name: %q, parent: %q, index: %d, version: %v }", t.arch, t.kind, t.name, t.parent, t.index, t.version.String())
+}
+
+type entryList []entryTuple
```
[the remark this answers is not in the document]
Comment (grokspawn) [reply alpha/veneer/semver/types.go]: Nah, we can totally do without it.  In one of the interim attempts for this feature, the entryList was a complex structure, so abstracting it made sense.  This was just left in as an available 'rough edge' to sand off.

**Pull request merged**